pkg/replay.go

```diff
@@ -20,97 +20,118 @@ package ripley
 
 import (
 	"bufio"
-	"encoding/json"
+	"container/heap"
 	"fmt"
 	"os"
 	"sync"
 	"time"
+	"unsafe"
+
+	"github.com/VictoriaMetrics/metrics"
 )
 
-func Replay(phasesStr string, silent, dryRun bool, timeout int, strict bool, numWorkers int) int {
+type Options struct {
+	Pace                string
+	Silent              bool
+	SilentHttpError     bool
+	DryRun              bool
+	Timeout             int
+	TimeoutConnection   int
+	Strict              bool
+	Memprofile          string
+	NumWorkers          int
+	PrintStat           bool
+	MetricsServerEnable bool
+	MetricsServerAddr   string
+	NlongestPrint       bool
+	NlongestResults     int
+}
+
+// Ensures we have handled all HTTP request results before exiting
+var waitGroupResults sync.WaitGroup
+
+func Replay(opts *Options) int {
 	// Default exit code
 	var exitCode int = 0
-	// Ensures we have handled all HTTP request results before exiting
-	var waitGroup sync.WaitGroup
 
 	// Send requests for the HTTP client workers to pick up on this channel
-	requests := make(chan *request)
+	requests := make(chan *Request, opts.NumWorkers)
 	defer close(requests)
 
 	// HTTP client workers will send their results on this channel
-	results := make(chan *Result)
+	results := make(chan *Result, opts.NumWorkers)
 	defer close(results)
 
 	// The pacer controls the rate of replay
-	pacer, err := newPacer(phasesStr)
-
+	pacer, err := newPacer(opts.Pace)
 	if err != nil {
 		panic(err)
 	}
 
 	// Read request JSONL input from STDIN
-	scanner := bufio.NewScanner(os.Stdin)
+	reader := bufio.NewReaderSize(os.Stdin, 1024*1024)
+	scanner := bufio.NewScanner(reader)
 
 	// Start HTTP client goroutine pool
-	startClientWorkers(numWorkers, requests, results, dryRun, timeout)
+	startClientWorkers(opts, requests, results)
 	pacer.start()
 
 	for scanner.Scan() {
-		req, err := unmarshalRequest(scanner.Bytes())
+		if pacer.done {
+			break
+		}
+		waitGroupResults.Add(1)
```

Comment:
I think this reads cleaner if we put it on right before line 100 (`results <- res`)

Reply:
`waitGroupResults.Add` is necessary there because the subsequent [code]((https://github.com/loveholidays/ripley/pull/8/files/f3c5427c08535caa6e9f0e8f7d4cf7b8515edea4#diff-eb03cca6be7ab66a2f8c8f0f9b35c1422187587153b292733eee7cb7d65f1d93R100)) sends to the results channel that runs [waitGroupResults.Done](https://github.com/loveholidays/ripley/pull/8/files/f3c5427c08535caa6e9f0e8f7d4cf7b8515edea4#diff-cd1bb4050bb9293a9065a48efa303e33d0ea88da5130e195f9dece09a6ce8be1R92) function. If you move it after, an error may occur.

I also moved to the beginning - the `if pacer.done` condition to stop read requests as quickly as possible. However, I don't mind to move it back.